image/qcow2/qcow2.go

```diff
@@ -959,6 +959,103 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
 	return l, err
 }
 
+// clusterStatus returns an extent describing a single cluster. off must be aligned to
+// cluster size.
+func (img *Qcow2) clusterStatus(off int64) (image.Extent, error) {
+	status := image.Extent{Start: off, Length: int64(img.clusterSize)}
+	var cm clusterMeta
+	if err := img.getClusterMeta(off, &cm); err != nil {
+		return status, err
+	}
+	if !cm.Allocated {
+		if img.backingImage == nil {
+			// Cluster must read as zeros.
+			status.Zero = true
+			return status, nil
+		}
+		// Cluster may present in the backing file.
+		return img.backingImage.Next(off, int64(img.clusterSize))
+	}
+	// Cluster present in this image.
+	status.Allocated = true
+	desc := cm.L2Entry.clusterDescriptor()
+	if cm.Compressed {
+		status.Compressed = true
+	} else {
+		standardDesc := standardClusterDescriptor(desc)
+		if img.extendedL2() {
+			// TODO: report entire cluster as data until we understand better the
+			// semantics of extended l2 clusters. This is correct but inefficient.
+			status.Zero = false
+		} else {
+			status.Zero = standardDesc.allZero()
+		}
+	}
+	return status, nil
+}
+
+// Return true if extents have the same status.
+func sameStatus(a, b image.Extent) bool {
+	return a.Allocated == b.Allocated && a.Zero == b.Zero && a.Compressed == b.Compressed
+}
+
+// Next returns the next extent in this image. An extent describes one or more
+// clusters having the same status.
+func (img *Qcow2) Next(start, length int64) (image.Extent, error) {
+	var current image.Extent
+
+	if img.errUnreadable != nil {
+		return current, img.errUnreadable
+	}
+	if img.clusterSize == 0 {
+		return current, errors.New("cluster size cannot be 0")
+	}
+
+	// Do not allow requests after the end of the image. This is likely a bug in
+	// the caller code, so better to fail fast and keep the code simpler.
```

Comment:
This does not work for the case we test:
```
base image (2g) <- top image (4g)
```

Read from child from offset 2g to 4g should return unallocated cluster - we get the same result when reading after the end of the image.